### ISSUE TITLE
CookieAgent to add support for HTTP cookie handling

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4922]
|<img alt="gabrtv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gabrtv reported|
|-|-|
|Trac ID|trac#4922|
|Type|enhancement|
|Created|2011-03-04 03:31:10Z|
|Branch|https://github.com/twisted/twisted/tree/cookie-agent-4922|

This patch extents the twisted.web.client.Agent to add support for handling HTTP cookies.  The new CookieLibAgent requires a cookieJar that implements the cookielib.CookieJar interface.  Cookies are then automatically set and extracted during Agent request processing.

At the time of writing, this patch was built atop [#3420](https://github.com/twisted/twisted/issues/3420) (Agent Persistent Connections).

This submission does not yet have tests.  However it is performing relatively well inside a larger unit test infrastructure.  I will try to add tests as my schedule permits, but I wanted to contribute the patch now in case others might find it useful.

Attachments:

* [cookieagent.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1eb/1ebfbdaf620003b846d51c9e9bbd90a43d363ee7/65c802b2d18433bf691782a6ea9bcbce3824d8c1.patch) (5106 bytes) - added by gabrtv on 2011-03-04 03:31:56Z - CookieAgent implementation
* [cookieagent_composed.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1eb/1ebfbdaf620003b846d51c9e9bbd90a43d363ee7/445bcfb2fd382cce86f3faf74abe3379393361c1.patch) (5136 bytes) - added by gabrtv on 2011-03-11 06:11:15Z - CookieAgent, composed
* [cookieagent_composed2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1eb/1ebfbdaf620003b846d51c9e9bbd90a43d363ee7/a2b53a749f6c7082a3c3d104c8f16b97c223ada2.patch) (4168 bytes) - added by gabrtv on 2011-03-11 06:22:44Z - composed, minus dependency on persistent connections

<details><summary>Searchable metadata</summary>

```
trac-id__4922 4922
type__enhancement enhancement
reporter__gabrtv gabrtv
priority__normal normal
milestone__ 
branch__branches_cookie_agent_4922 branches/cookie-agent-4922
branch_author__Gabriel_Monroy__jonathanj Gabriel Monroy, jonathanj
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1299209470000000 1299209470000000
changetime__1305467180000000 1305467180000000
version__None None
owner__jonathanj jonathanj
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="gabrtv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|gabrtv commented|
|-|-|

I should note that CookieAgent supports FileCookieJars which presumably perform some blocking file i/o (such as cookielib.MozillaCookieJar).  Not sure if that's a problem or not.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to gabrtv|
|-|-|

Thanks for working on this gabrtv.  It will be great to have cookie support for `Agent` in Twisted.

The primary comment I have about this patch is that it would be simpler and more reusable if it _composed_ a cookie-aware `Agent`-alike with `Agent` instead of _subclassing_ `Agent`.  This will also remove the dependency on the persistent connections branch.

**Comment by twisted-trac** [comment]
|<img alt="gabrtv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|gabrtv commented|
|-|-|

CookieAgent patch is now composed, and should no longer have a dependency on agent persistent connections branch.

**Comment by twisted-trac** [comment]
|<img alt="gabrtv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|gabrtv removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

This is meant to be in review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to gabrtv|
|-|-|

Thanks for the patch. The review:

 * You need to add tests.

 * The coding standard mentions 2 blank lines between methods, 3 between classes.

 * `def __init__(self, agent, cookieJar, *args, **kwargs):` It seems the extra arguments are unused.

 * It seems that the docstring of the request methods is an exact copy of the one of Agent.request. You should rather refer to Agent.request, and mentions the specific bits.

 * There may be a design problem with the way you manage the cookie jar: an Agent is stateless, so you should be able to issue as many requests as you want against one. The jar is "global" to the CookieAgent, so is modified after each response.

 * You set the cookie header directly on the header object. First, I think you forgot that headers is an optional argument, so it's necessarily existing. Then, I think we try not to modify the passed argument, so you should make a copy of the headers before.

 * It would be nice if you could define a fake urllib2 Request class, as you did for the response.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_9"></a>|@jonathanj commented|
|-|-|

(In [31792]) Branching to 'cookie-agent-4922'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_10"></a>|@jonathanj commented|
|-|-|

(In [31805]) Apply cookieagent_composed2.patch from gabrtv. Address review commentary. Refs [#4922](https://github.com/twisted/twisted/issues/4922).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj removed owner|
|-|-|

Replying to [therve](#note_8):
>  * There may be a design problem with the way you manage the cookie jar: an Agent is stateless, so you should be able to issue as many requests as you want against one. The jar is "global" to the CookieAgent, so is modified after each response.

As we discussed on IRC, I think that this is probably how you want `CookieAgent` to work. If you need separate cookie jars use separate cookie jars and agents; I added a sentence in the `CookieAgent` docstring to clarify that the cookie jar will be mutated.

Addressed all your other review points.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set owner to @jonathanj|
|-|-|

Looks great. Some comments:

 * Please refer to the new IResponse interface instead of the Response class in docstrings.
 * The NEWS fragment should be slightly rephrased I think. Maybe "A new HTTP cookie-aware Twisted Web Agent wrapper is included in twisted.web.client.CookieAgent".

Please merge once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_13"></a>|@jonathanj commented|
|-|-|

(In [31863]) Refer to twisted.web.iweb.IResponse instead of twisted.web.client.Response, refs [#4922](https://github.com/twisted/twisted/issues/4922).

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_14"></a>|@jonathanj set status to `closed`|
|-|-|

(In [31868]) Merge cookies-agent-4922.

Author: gabrtv, jonathanj
Reviewer: exarkun, therve
Fixes: [#4922](https://github.com/twisted/twisted/issues/4922)

Introduce twisted.web.client.CookieAgent, an HTTP cookie-aware Agent wrapper, that will read and write cookies to a cookielib.CookieJar instance.
